### DISCUSSION TITLE
Source Plugin for Notion

### DISCUSSION THREAD

**Discussion by LekoArts** [gatsbyjs/gatsby#31609] [Community]
Hello!

This might seem unusual but I figured it's the best way to contact everyone and start a discussion. Potentially this could also be moved to a private conversation but I don't see any harm in beginning this in the open.

## Premise

[Notion](https://www.notion.so/) currently has their API in a Beta and everyone is already very excited about it. For Gatsby it'll be most probably a very popular plugin/source to have and to create websites with. Thus it's critical that the plugin for it is well-written (see [Creating A Source Plugin](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/creating-a-source-plugin/)), easy to find and maintained.

One problem I see arising at the moment (which is also an opportunity) is that multiple people start publishing different plugins for the same source. Once everyone can use Notion as an API people will eventually go to gatsbyjs.com/plugins, search for "Notion" and find 5-6 different plugins. It'll be hard for them to see which one they use. Once they decide to use one plugin they might see that another one has feature X they need, but not feature Y that the current plugin has.

### Plugins

- https://github.com/cstrnt/gatsby-source-notion from @cstrnt
- https://github.com/pcarion/gatsby-source-notionso from @pcarion
- https://github.com/thedaviddias/gatsby-source-notion-article from @thedaviddias
- https://github.com/danielkvist/gatsby-source-notion-db from @danielkvist
- https://github.com/wonism/gatsby-source-notion-contents from @wonism
- https://github.com/orlowdev/gatsby-source-notion-api from @orlowdev
- https://github.com/alvis/gatsby-source-notion from @alvis

## Solution

I don't know why those different plugins were created and how they are different from each other -- but I _really_ wish that y'all would join forces on one plugin and make that one really great!

It'll greatly benefit the Gatsby community as:

- No decision fatigue for potential users
- Lower risk of not having time to maintain the plugin as multiple people look after it
- Easier for us (Gatsby team) to potentially help out with improvements as we have one place to direct feedback to

For me personally this is also something I love about open source: Collaborating with others on a project.

**Of course**, I can't force y'all to join projects into one repository. You may have reasons to not agree to this initially (e.g. wanting to have authority, wanting "recognition" for your plugin, etc.) but those things are solvable. Speaking from personal experience it's more often than not fun to work on something like this together.

Create a new github organization & create a repository there, use the npm package name from @alvis, add all people as contributors and get rolling 👍

**Comment by pragmaticpat** [comment]
This is definitely the type of behavior we like to see - multiple folks with similar goals coming together to co-create something _great_. Thanks for kicking this off @LekoArts !

**Comment by orlowdev** [comment]
Hi everyone,

Just wanted to highlight that some of the packages do not use the official Notion API released recently - `gatsby-source-notion` and `gatsby-source-notionso` seem to be on the list, and will probably want to preserve that unique behavior. I also used one of them myself for quite some time before Notion API release.

As for others, including mine (`gatsby-source-notion-api`), looks like they are using the official API.

I’d be glad to collaborate, moving/renaming my package is absolutely fine by me as long as I’m listed as a contributor. I don’t **own** this code after all. And I hope it’ll be a great way to learn for me. 👍

@LekoArts guidance, organisation and support would be very welcome! 🙂

**Reply by orlowdev** [reply]
My personal intent was to make a source plugin that will provide output compatible with `MarkdownRemark` and `MDX` - page content as markdown, and page props as frontmatter.

Currently wading through issues and current limitations of the API because of its very early stage.

**Reply by orlowdev** [reply]
Update, `gatsby-source-notion-contents` isn't based on the official API as well.

**Reply by LekoArts** [reply]
> gatsby-source-notion and gatsby-source-notionso seem to be on the list, and will probably want to preserve that unique behavior [+ gatsby-source-notion-contents]

Most often the correct approach is to use official APIs & packages so I expect those packages to be obsolete soon

So then the quest is to combine the plugins that use the current beta API.

**Comment by thedaviddias** [comment]
Hey @LekoArts and everyone!

Thank you so much for providing us with a way to all come together. I'm the creator of the plugin "Gatsby-Source-Notion-Article". Just to give a little more context, as @orlowdev, most of the source plugins mentioned above don't use the official Notion API.

When I started working on my plugin, no one had created any plugin for Gatsby yet using the new Notion API. I decided to take a different approach than a possible "official" plugin and this is the reason why I used "-article" at the end. Based on some people's asks, I decide to create a plugin that would bring an article created on Notion in Gatsby, nothing more.

I would be absolutely up to revisit the plugin and kill it if it adds some confusion or is not something that people are looking for. 
I agree with all the points @LekoArts mentioned, better to be more than one person on a project. I know how painful it could be to be alone, especially in a project that a lot of people use but not a lot want to invest time to help with.

**Reply by LekoArts** [reply]
Thanks for giving context on this! I think the functionality you have can be moved into a more general/generic plugin.

**Comment by orlowdev** [comment]
Hi y'all, I recently published 0.8.1 of `gatsby-source-notion-api` and migrated my blog at [https://orlow.dev](https://orlow.dev) to Notion as a CMS. Whatever works on my blog would work for anyone else out there. Technically, the PoC has proven to be successful. As far as current Notion API limitations are concerned, of course. Now I'm going to do some refactoring and I'd like to ask for some links on how I should organize the code to better suit the code standards.

I also have a feeling that my plugin acts both as a **source** and a **transformer** because I pull data from Notion and then transform it to be edible by MarkdownRemark and MDX. Should this be decoupled?

**Reply by LekoArts** [reply]
As long as the end result of the source plugin are nodes that have the `mimeType` of `text/markdown` other transformers and plugins will be able to pick this up. See: https://www.gatsbyjs.com/docs/creating-a-source-plugin/#setting-media-and-mime-types

**Reply by orlowdev** [reply]
Thanks. That is correct, the source plugin I made creates nodes with this MIME type. Notion pages are both available via `markdownRemark` and `notion`. I didn't test with MDX yet, but I think it should be the same as with MarkdownRemark, since they both look at the MIME type, from what I know.

**Comment by LekoArts** [comment]
Pinging @cstrnt - @osmanmesutozcan - @pcarion - @danielkvist - @wonism again :)

If that doesn't help my plan of action would be:

1) Reach out to npm support about getting a grip of https://www.npmjs.com/package/gatsby-source-notion
2) See if I can reach the other people above via other means (twitter, email, etc)

**Comment by alvis** [comment]
Hi everyone, it happens that I am also writing a notion source plugin for gatsby. (😓  didn't see this discussion before)
https://github.com/alvis/gatsby-source-notion

Besides multi-database support, it also features a not-so-perfect preview mode similar to the sanity plugin.
Well. It's not as quick as sanity as Notion hasn't got a webhook yet, but it does the job.

Check it out and let me know what you guys think? :)

**Reply by LekoArts** [reply]
It seems like you got hold of the npm package name? At least one problem solved with this then as in someone active has access to this now.

**Reply by orlowdev** [reply]
I'd be glad to join

**Reply by LekoArts** [reply]
I'll try to find the email of everyone now and send out an email to everyone individually so that they at least see this here and can comment if they're interested at least.

**Reply by alvis** [reply]
Yes. @osmanmesutozcan was so kind to pass the package right earlier so we can make something great here.
& sure. I'm in.

**Comment by cstrnt** [comment]
Hello everyone, I'm interested in helping with this :)

**Comment by dnlzrgz** [comment]
Hello everyone!

I am sorry I have not been able to participate in this discussion before. I'm the developer of the [`gatsby-source-notion-db`](https://www.npmjs.com/package/gatsby-source-notion-db), which in my case I am going to archive. For personal reasons I don't think I will have time, but I will try to participate at least occasionally in the development of `gatsby-source-notion`.
